temporalio/client.py

```diff
@@ -0,0 +1,935 @@
+
+
+class Client:
+    """Client for accessing Temporal.
+
+    Most users will use :py:meth:`connect` to create a client. The
+    :py:attr:`service` property provides access to a raw gRPC client. To create
+    another client, like for a different namespace, :py:func:`Client` may be
+    directly instantiated with a :py:attr:`service` of another.
+    """
+
+    @staticmethod
+    async def connect(
+        target_url: str,
+        *,
+        namespace: str = "default",
+        identity: str = f"{os.getpid()}@{socket.gethostname()}",
+        data_converter: temporalio.converter.DataConverter = temporalio.converter.default(),
+        interceptors: Iterable["Interceptor"] = [],
+        workflow_query_reject_condition: Optional[WorkflowQueryRejectCondition] = None,
+    ) -> "Client":
+        """Connect to a Temporal server.
+
+        Args:
+            target_url: URL for the Temporal server. For local development, this
+                is often "http://localhost:7233".
+            namespace: Namespace to use for client calls.
+            identity: Identity to use for client calls.
+            data_converter: Data converter to use for all data conversions
+                to/from payloads.
+            interceptors: Set of interceptors that are chained together to allow
+                intercepting of client calls. The earlier interceptors wrap the
+                later ones.
+            workflow_query_reject_condition: When to reject a query.
+        """
+        return Client(
+            await temporalio.workflow_service.WorkflowService.connect(target_url),
+            namespace=namespace,
+            identity=identity,
+            data_converter=data_converter,
+            interceptors=interceptors,
+            workflow_query_reject_condition=workflow_query_reject_condition,
+        )
+
+    def __init__(
+        self,
+        service: temporalio.workflow_service.WorkflowService,
+        *,
+        namespace: str = "default",
+        identity: str = f"{os.getpid()}@{socket.gethostname()}",
+        data_converter: temporalio.converter.DataConverter = temporalio.converter.default(),
+        interceptors: Iterable["Interceptor"] = [],
+        workflow_query_reject_condition: Optional[WorkflowQueryRejectCondition] = None,
+    ):
+        """Create a Temporal client from a workflow service.
+
+        See :py:meth:`connect` for details on the parameters.
+        """
+        self._service = service
+        self._namespace = namespace
+        self._identity = identity
+        self._data_converter = data_converter
+        self._interceptors = interceptors
+        self._workflow_query_reject_condition = workflow_query_reject_condition
+
+        # Iterate over interceptors in reverse building the impl
+        self._impl: OutboundInterceptor = _ClientImpl(self)
+        for interceptor in reversed(list(interceptors)):
+            self._impl = interceptor.intercept_client(self._impl)
```

Comment:
Hmmm... in TS we instantiate interceptors per request https://github.com/temporalio/sdk-typescript/blob/67526d3a9effaea51d9aac3573ba2f8c9d3e1bab/packages/client/src/workflow-client.ts#L225.

AFAIR this was copied from the Java implementation but I don't remember the reasoning, I'll probably change TS to work this way too.

Reply:
> Hmmm... in TS we instantiate interceptors per request

In Go we do not. We allow people to have per-interceptor instead of per RPC call state.

Reply:
I would change TS here.